apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/coordinator-tools.ts

```diff
@@ -113,23 +146,72 @@ export const coordinatorToolDefinitions: Record<
       "Terminate the research task, because it cannot be completed with the provided tools.",
     inputSchema: {
       type: "object",
-      properties: {},
-      required: [],
+      properties: {
+        explanation: explanationDefinition,
+      },
+      required: ["explanation"],
     },
   },
   updatePlan: {
     name: "updatePlan",
-    description:
-      "Update the plan for the research task. You should call this alongside other tool calls to progress towards completing the task.",
+    description: dedent(`
+      Update the plan for the research task.
+      You can call this alongside other tool calls to progress towards completing the task.
+    `),
     inputSchema: {
       type: "object",
       properties: {
+        explanation: {
+          type: "string",
+          description: dedent(`
+            An explanation of why the plan needs to be updated, and
+            how the updated plan aligns with the task.
+          `),
+        },
         plan: {
           type: "string",
           description: "The updated plan for the research task.",
         },
       },
-      required: ["plan"],
+      required: ["plan", "explanation"],
+    },
+  },
+  proposeAndSubmitLink: {
+    name: "proposeAndSubmitLink",
+    description: dedent(`
+      Propose and submit a link between two entities.
+      The source or target entity can be:
+        - a previously submitted proposed entity
+        - an existing entity
+
+      If you want to use a proposed entity as the source or target entity, you must first
+        submit the proposed entity using the "submitProposedEntities" tool.
+        
+      Proposing a link will also submit it as a proposed link entity for the research task.
```

Comment:
> Can the model not submit this alongside a call to submit the source/target entity? How does it operate in practice? Maybe we can just do it automatically on its behalf.

It would probably be the least brittle if we just handle submitting source/entity proposed entities as part of the tool call, have made this change in https://github.com/hashintel/hash/pull/4346/commits/a859ae7abf0437ed789bae3b72c95cea958027f9 and also applied some of your suggestion to simplify the tool's description.

> There is also no ability to include properties here – they may still be inferrable from the content provided by the inference agent (e.g. an entity's description might mention something which provides a property value for a link). See how this works in practice in follow-up, but I imagine we'll probably end up just wanting to give this agent much the same tools as the inference agent to create entities and link entities in full with knowledge of their schemas (it'll need this to do deduplication/merging as well)

I agree, I left this as a `@todo` because I thought validation of the properties wasn't yet possible via the Graph API (at a later point I was corrected in an internal thread). I'm not sure how to best test this in its current form, as the coordinator agent doesn't really have much information to work with when proposing these links, but if we have use-cases later on where the coordinator (or worker agent) may be able to create properties on a link we can add this functionality then.